text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,833 @@
+
+The type of the expression `become <call>` is `!` (the never type, see [here](https://doc.rust-lang.org/std/primitive.never.html)). This is consistent with other control flow constructs such as `return`, which also have the type of `!`.
+
+Note that as `become` is a keyword reserved for exactly the use-case described in this RFC there is no backward-compatibility break. This RFC only specifies the use of `become` inside of functions and instead leaves usage outside of functions unspecfied for use by other features.
+
+This feature will have interactions with other features that depend on stack frames, for example, debugging and backtraces. See [drawbacks](#drawbacks) for further discussion.
+
+See below for specifics on interactions with other features.
+
+## Coercions of the Tail Called Function's Return Type
+[return-type-coercion]: #return-type-coercion
+
+All coercions that do any work (like deref coercion, unsize coercion, etc) are prohibited.
+Lifetime-shortening coercions (`&'static T` -> `&'a T`) are allowed but will be checked by the borrow checker.
+
+Reference/pointer coercions of the return type are **not** supported to minimize implementation effort. Though, coercions which don't change the pointee (`&mut T -> &T`, `*mut T -> *const T`, `&T -> *const T`, `&mut T -> *mut T`) could be added in the future.
+
+Never-to-any coercions (`! -> T`) of the return type are **not** supported to minimize implementation effort. They are difficult to implement and require backend support. To be clear, this only concerns functions that have the never return type like the following example:
+
+```rust
+fn never() -> ! {
+    loop {}
+}
+
+fn tail_call_never_type() -> usize {
+    become never(); //~ error: mismatched types
+}
+```
+
+## Closures
+[closures]: #closures
+
+Tail calling closures _and_ tail calling _from_ closures is **not** allowed.
+This is due to the high implementation effort, see below, this restriction can be lifted by a future RFC.
+
+Closures use the `rust-call` unstable calling convention, which would need to be adapted to guarantee TCE.
+Additionally, any closure that has captures would need special handling, since the captures would currently be dropped before the tail call.
+
+## Variadic functions using `c_variadic`
+
+Tail calling [variadic functions](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html) _and_ tail calling _from_ variadic functions is **not** allowed.
+As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but supported can be added with a future RFC.
```

Comment:
```suggestion
As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but support can be added with a future RFC.
```